RE: The setting to change the Language of the game to Any language

JackTriton opened the pull request:
Hi, it's been a long time
This PR is redo of all of the functions I made almost 5 years ago

## Description
With this PR, you can play OOTR with different languages or event texts / custom texts as well  

----
## Addition
### Files
* **language (folder)**: Consists every language you can play with  
Also, you can add another language file just by drag and drop the file with same structure  
For how to create and what file is included, checkout `language / README.md`
* **Language.py**: allows to manage language files (json, bin, ia4)
----
### Additional Control Codes for wide characters  

* All of the additional control codes are placed the code + 0x8700 (for example, silver rupee count is placed 0xF0 + 0x8700 = 0x87F0)  
For the character to use, check out `Messages.py`
----
### Text alignment
* You can change the alignment of the text to **Left**, **Center** and **Right**

----

## How to Change the language?
You can change the language by changing `Main Rules >> Language >> Language Selection`

----
## Tips for language creation
If you want to add language with characters not included in vanilla game (for example Spanish has `ñ`), do the same thing as other images: decompile the original game, replace some files with new one and recompile / get diff with `get_diff.py`  

----
## Notes
Currently, besides English which is already implemented from beginning, it allows you to play with Japanese as well  
Why Japanese? well, it's because Japanese is the other language that NTSC rom has and one of language that uses wide characters instead and also because I'm Japanese  

Just because I didn't implemented other languages such as German, French or Chinese (which is PAL and iQue version has), that doesn't mean you can't play with those languages or other languages that only got the translated version by using mod / patch or even that never got the version  

Not only that, you can create version for events, competition or even speedruns as well (like shortening all of the texts, and speed up everything)

## Screen shots
![Setting View](https://raw.githubusercontent.com/JackTriton/OOTR-Langfig/main/Setting_view.png "Setting View")

![Title for Japanese](https://raw.githubusercontent.com/JackTriton/OOTR-Langfig/main/Ingame_Japanese.png "Title Screen for OOTR Japanese")

![Title for English](https://raw.githubusercontent.com/JackTriton/OOTR-Langfig/main/Title_English.png "Title Screen for OOTR English")

fenhl reviewed (commented):
Preliminary review. I've already mentioned some of this on Discord but am repeating it here for reference.

* The hint area type being language-dependent is going to be an issue for parts of the codebase that don't directly interact with text and therefore don't care about the language. Currently, if I understand the code correctly, these just use an English-language version of the hint area type. This will cause issues if such a hint area value does end up in language-specific code and causes text to be generated in English instead of the correct language. I'd separate the language information from the hint area by passing it as a parameter to methods that generate text.
* Some of the variable names and magic strings (e.g. magic strings called `o`, `wiw`, and `cbf`; single-letter variable names in `Language.format_from_text` and `Patches.create_fake_name`) are difficult to understand at a first glance. I would recommend replacing them with longer, more descriptive names.
* There seem to be some pieces of logic that check if the language is English and use the logic for Japanese in the `else` case. The level of Japanese proficiency among maintainers and contributors is much lower than that of English, so for maintainability, English should be the fallback case.
* The PR adds `.DS_Store` files to the repo. These should be removed, and I recommend adding them to your `.git/info/exclude` to prevent accidentally adding them back in the future.
* The JSON file uses fixed-length (tuple-like) arrays in some places. To make it easier to read, I suggest replacing these with objects (`{}`) with descriptive property names.

I will take a closer look at specific parts of the code and go into more detail in a later review, once CI is green.

JackTriton commented:
Most of the system reworked, deleted .DS_Store, add some descriptions to the `property_build.py` and changed some of the values to have clearer infos

cjohnson57 commented on SettingsList.py:
```diff
@@ -67,6 +76,16 @@ class SettingInfos:
 
+    language_selection = Combobox(
+        gui_text       = 'Language Selection',
+        default        = 'japanese',
```
Two issues:

1) Because the default is not capitalized, the box is blank by default, as the options grabbed from the folder names are capitalized.

2) Default should be English.

Change to `English`

cjohnson57 commented on SettingsList.py:
```diff
@@ -67,6 +76,16 @@ class SettingInfos:
         },
     )
 
+    language_selection = Combobox(
```
This should be moved down to be in logical order like the rest of the settings. So it should be after the `logic_rules` setting in this file.

cjohnson57 commented on SettingsList.py:
```diff
@@ -67,6 +76,16 @@ class SettingInfos:
+        choices        = get_language(),
+        gui_tooltip    = '''\
+            Language sets the one that you use on the game itself
+            Some languages requires using NTSC rom instead
```
Some grammar and clarity issues here.

Recommended change: "Sets the language used within the game." (make sure to include the period)

Remove the second line, the randomizer already requires NTSC so it's unneeded. If future languages require different roms, we can specify that here when they're added.

cjohnson57 reviewed (changes requested):
Some pretty major (but easily fixable) issues with the setting definition itself, adding this separately before I review the rest of it.

cjohnson57 commented:
Got different errors when trying to generate in English and Japanese.

English:
```
<Dungeon.Dungeon object at 0x000001EEE96AD480>
Traceback (most recent call last):
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\OoTRandomizer.py", line 57, in start
    main(settings)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 57, in main
    patch_and_output(settings, spoiler, rom)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 375, in patch_and_output
    patch_cosmetics_log = prepare_rom(spoiler, world, rom, settings, rng_state, restore_rom)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 214, in prepare_rom
    patch_rom(spoiler, world, rom)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Patches.py", line 1849, in patch_rom
    "dungeon_reward": world.language.hintTable[dungeon][1],
KeyError: <Dungeon.Dungeon object at 0x000001EEE96AD480>
```

Seems to be because the dungeon name has a control code in it. Also, pretty sure the `the` should be dropped and it should just be Deku Tree...
<img width="956" height="77" alt="image" src="https://github.com/user-attachments/assets/544eeb9f-ff10-4929-92a5-42af11c19937" />
<img width="1079" height="373" alt="image" src="https://github.com/user-attachments/assets/fdabfb8e-7beb-4703-9d68-d22b36a2f092" />


Japanese:
```
from_bytes() missing required argument 'byteorder' (pos 2)
Traceback (most recent call last):
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\OoTRandomizer.py", line 57, in start
    main(settings)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 57, in main
    patch_and_output(settings, spoiler, rom)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 375, in patch_and_output
    patch_cosmetics_log = prepare_rom(spoiler, world, rom, settings, rng_state, restore_rom)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 214, in prepare_rom
    patch_rom(spoiler, world, rom)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Patches.py", line 1176, in patch_rom
    update_message_by_id(messages, 0x305C, lang.format_from_id("PATCH_TEXTS.claim"), lang)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Messages.py", line 748, in update_message_by_id
    text = line_wrap(text, lang.base, align=lang.lang_property["align_text"])
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\TextBox.py", line 51, in line_wrap
    text_codes = Messages.parse_control_codes(text, lang)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Messages.py", line 345, in parse_control_codes
    text_bytes = encode_text_string(text) if lang else encode_text_string_jp(text)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Messages.py", line 295, in encode_text_string_jp
    result.append(int.from_bytes(h.encode("cp932")))
TypeError: from_bytes() missing required argument 'byteorder' (pos 2)
```

Adding `byteorder="big"` to the function call fixed it, but I'm not actually sure why that was necessary when the function definition already sets it to big by default? `result.append(int.from_bytes(h.encode("cp932"), byteorder="big"))`

```python
def from_bytes(
            cls,
            bytes: Iterable[SupportsIndex] | SupportsBytes | ReadableBuffer,
            byteorder: Literal["little", "big"] = "big",
            *,
            signed: bool = False,
        )
```

@flagrama any ideas here?

Anyway, fixing that issue ran into the same issue English did with the dungeon name.
<img width="945" height="104" alt="image" src="https://github.com/user-attachments/assets/837d9303-f51f-4965-8353-9784129a88f9" />

cjohnson57 commented on HintList.py:
```diff
@@ -48,8 +49,9 @@ def __init__(self, name: str, locations: list[str]) -> None:
-def get_hint(name: str, clearer_hint: bool = False) -> Hint:
-    text_options, clear_text, hint_type = hintTable[name]
+def get_hint(name: str, lang: Language, clearer_hint: bool = False) -> Hint:
+    text_options, clear_text, _ = lang.hintTable[name]
+    _, _, hint_type = hintTable[name]
```
Are these underscores needed?

JackTriton replied:
No, I think it would be better if I use `hint_type = hintTable[name][2]` for now.
The last underscore of `lang.hintTable[name]` would be for gender value so I'll just leave it as underscore for future cases

cjohnson57 commented on Language.py:
```diff
@@ -0,0 +1,91 @@
+    def format_from_id(self, id: str, external: dict = None):
+        keys = id.split('.')
+        if len(keys)>1:
+            key=keys.pop(0)
```
For consistent styling there should be spaces before and after the equals sign, like `key = keys.pop(0)` (also applies to several other places in this file)

JackTriton commented:
> Got different errors when trying to generate in English and Japanese.
> 
> English:
> 
> ```
> <Dungeon.Dungeon object at 0x000001EEE96AD480>
> Traceback (most recent call last):
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\OoTRandomizer.py", line 57, in start
>     main(settings)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 57, in main
>     patch_and_output(settings, spoiler, rom)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 375, in patch_and_output
>     patch_cosmetics_log = prepare_rom(spoiler, world, rom, settings, rng_state, restore_rom)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 214, in prepare_rom
>     patch_rom(spoiler, world, rom)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Patches.py", line 1849, in patch_rom
>     "dungeon_reward": world.language.hintTable[dungeon][1],
> KeyError: <Dungeon.Dungeon object at 0x000001EEE96AD480>
> ```
> 
> Seems to be because the dungeon name has a control code in it. Also, pretty sure the `the` should be dropped and it should just be Deku Tree... <img alt="image" width="956" height="77" src="https://private-user-images.githubusercontent.com/30701749/476320828-544eeb9f-ff10-4929-92a5-42af11c19937.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ3ODQ3NDksIm5iZiI6MTc1NDc4NDQ0OSwicGF0aCI6Ii8zMDcwMTc0OS80NzYzMjA4MjgtNTQ0ZWViOWYtZmYxMC00OTI5LTkyYTUtNDJhZjExYzE5OTM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODEwVDAwMDcyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhZjA0N2UxODY3ZmVjOWQ4ODM4NDM2MGJjZTlmMzNmODY4ZTU3OWJiYjVkYTRmODU4OGI5M2E4ZGYzZmU4ODQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fxaQSFjcHJpxrizSvxlK-hdQ2c-K2lhe-NnOV4C4vng"> <img alt="image" width="1079" height="373" src="https://private-user-images.githubusercontent.com/30701749/476321037-fdabfb8e-7beb-4703-9d68-d22b36a2f092.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ3ODQ3NDksIm5iZiI6MTc1NDc4NDQ0OSwicGF0aCI6Ii8zMDcwMTc0OS80NzYzMjEwMzctZmRhYmZiOGUtN2JlYi00NzAzLTlkNjgtZDIyYjM2YTJmMDkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODEwVDAwMDcyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhNzNlZTFlNzg4ZmVlNjc2ODkzNzkwZjg3NDExOWE0Y2Y5NTYzNDM1NWRlMGZmOTUzODY3OWJiYTc5ZWZhMDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Z5KBx9GspZFEhsiR21BcMhLRkIpPKWvkYP-Ab5OLuRM">
> 
> Japanese:
> 
> ```
> from_bytes() missing required argument 'byteorder' (pos 2)
> Traceback (most recent call last):
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\OoTRandomizer.py", line 57, in start
>     main(settings)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 57, in main
>     patch_and_output(settings, spoiler, rom)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 375, in patch_and_output
>     patch_cosmetics_log = prepare_rom(spoiler, world, rom, settings, rng_state, restore_rom)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 214, in prepare_rom
>     patch_rom(spoiler, world, rom)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Patches.py", line 1176, in patch_rom
>     update_message_by_id(messages, 0x305C, lang.format_from_id("PATCH_TEXTS.claim"), lang)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Messages.py", line 748, in update_message_by_id
>     text = line_wrap(text, lang.base, align=lang.lang_property["align_text"])
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\TextBox.py", line 51, in line_wrap
>     text_codes = Messages.parse_control_codes(text, lang)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Messages.py", line 345, in parse_control_codes
>     text_bytes = encode_text_string(text) if lang else encode_text_string_jp(text)
>   File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Messages.py", line 295, in encode_text_string_jp
>     result.append(int.from_bytes(h.encode("cp932")))
> TypeError: from_bytes() missing required argument 'byteorder' (pos 2)
> ```
> 
> Adding `byteorder="big"` to the function call fixed it, but I'm not actually sure why that was necessary when the function definition already sets it to big by default? `result.append(int.from_bytes(h.encode("cp932"), byteorder="big"))`
> 
> ```python
> def from_bytes(
>             cls,
>             bytes: Iterable[SupportsIndex] | SupportsBytes | ReadableBuffer,
>             byteorder: Literal["little", "big"] = "big",
>             *,
>             signed: bool = False,
>         )
> ```
> 
> @flagrama any ideas here?
> 
> Anyway, fixing that issue ran into the same issue English did with the dungeon name. <img alt="image" width="945" height="104" src="https://private-user-images.githubusercontent.com/30701749/476320961-837d9303-f51f-4965-8353-9784129a88f9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ3ODQ3NDksIm5iZiI6MTc1NDc4NDQ0OSwicGF0aCI6Ii8zMDcwMTc0OS80NzYzMjA5NjEtODM3ZDkzMDMtZjUxZi00OTY1LTgzNTMtOTc4NDEyOWE4OGY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODEwVDAwMDcyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmZGQ3NmQ5YzYzYjIzMGVlYTNkZThlNzg0MzBjM2E2ZjRkYzU3MjE1YzBmOGM5ZWQ1ZTRjZjZiOGQwMTFmYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JVwflW1Lw8oVw7oyE0xQ3BWMecmEyscSkCATrf4gK7g">

LangDev here, this is probably because of the given value is `dungeon` instead of `dungeon_reward`

cjohnson57 commented on TextBox.py:
```diff
@@ -30,9 +32,13 @@
+def line_wrap(text: str, lang: str, strip_existing_lines: bool = False, strip_existing_boxes: bool = False, replace_control_chars: bool = True, align: str = "Left"):
     # Replace stand-in characters with their actual control code.
-    if replace_control_chars:
+    lang = 0 if lang == "jp" else 1
```
Is the idea that even as more languages are added, this value (as well as `NORMAL_LINE_WIDTH` and `LINES_PER_BOX`) will be the same for all of them, except JP?

If so then I think it's fine as it is, otherwise I would recommend we make a dictionary that maps language to `NORMAL_LINE_WIDTH` etc.

Either way though, I think just calling it lang is a bit confusing. Also it should not be overwriting the lang variable passed to the function. Maybe `lang_index`?

JackTriton replied:
JP has its own issue with characters and lines.  
Most of characters in JP has the width of 16 pixels thus the presented `1801800` cannot be used
Lines are set to 3 instead of 4 as well
Idea of lang_index would be great so I'll just edit them as well

cjohnson57 replied:
Makes sense

cjohnson57 commented on TextBox.py:
```diff
@@ -123,61 +143,86 @@ def replace_bytes(match: re.Match) -> str:
 
 
-def calculate_width(words: list[list[TextCode]]):
+def calculate_width(words: list[list[TextCode]], lang: str|int):
     words_width = 0
+    lang= 1 if lang in ["en",1] else 0
+    CC=Messages.CONTROL_CODES if lang else Messages.CC_PARSE_JP
```
More issues with equal sign spacing, and I think having lang be a string or an int is very awkward. Adds more reason for my previous comment to not overwrite the `lang` variable, and just always pass it as a string.

cjohnson57 commented on TextBox.py:
```diff
@@ -42,14 +48,14 @@ def replace_bytes(match: re.Match) -> str:
     strip_codes = []
     if strip_existing_boxes:
-        strip_codes.append(0x04)
+        strip_codes.append([0x81A5,0x04][lang])
```
Commas should have a space after them (applies to many other places in this file)

cjohnson57 commented on Patches.py:
```diff
@@ -1163,43 +1167,43 @@ def calculate_traded_flags(world):
-        update_message_by_id(messages, 0x70f8, new_message)
+        new_message = lang.format_from_id("PATCH_TEXTS.brought_poe",{"poe": poe_points})
+        update_message_by_id(messages, 0x70f7, new_message, lang)
+        new_message = lang.format_from_id("PATCH_TEXTS.enough_poe",{"poe": poe_points})
```
Should be `enough_poes`, also more issues with missing spaces after commas

cjohnson57 commented on Messages.py:
```diff
@@ -1374,31 +1180,32 @@ def update_warp_song_text(messages: list[Message], world: World) -> None:
-        else:
-            destination_name = 'to a mysterious place'
-            color = COLOR_MAP['White']
+    if world.settings.logic_rules != "glitched": # Entrances not set on glitched logic so following code will error
```
This is outdated now with the new advanced logic. You'll probably need to pull the latest dev changes and merge/update instances like this.

cjohnson57 commented on Hints.py:
```diff
@@ -350,25 +344,38 @@ def get_simple_hint_no_prefix(item: Item) -> Hint:
     text = gossip_text.text
     colors = list(gossip_text.colors) if gossip_text.colors is not None else []
     color = 'White'
+    if lang.base=="jp":
```
More spacing issues

cjohnson57 reviewed (changes requested):
Here's my review of the rest of the code. Admittedly mostly nitpicks, but a few actual issues.

Thanks so much for all your work on this!

flagrama commented:
> Adding `byteorder="big"` to the function call fixed it, but I'm not actually sure why that was necessary when the function definition already sets it to big by default? `result.append(int.from_bytes(h.encode("cp932"), byteorder="big"))`

FYI, the default value is just a change made in python 3.11. If you are using 3.11 or newer, this error won't appear. If you are using anything older this error comes up. If you are using 3.11 or newer but you have your dev environment set up to ensure everything works as far back as 3.8 it may be doing something additional to cause that error to show up.

JackTriton commented:
> One question, what exactly is the purpose of the `property_build.py` file? It seems to mainly contain the English messages that are now formatted in the English folder's `property.json`. Is it to generate the `property.json`?

`property_build.py` creates `property.json` for the language
The default value would be the ones on English for compare / know what string to use

Also, the `PLANE_TEXTS` would be the ones for other texts to implement in game when the language is not on NTSC or any kind of variety

cjohnson57 commented:
Gotcha. Should that be "plain" texts?

JackTriton commented:
> Gotcha. Should that be "plain" texts?

Yes, texts that is not patched in any other kind of message would consider be plain texts
eg. the below message would consider be plain as well)
```
0x0308
-----------------------
Death Mountain
No passage without a
Royal Decree!
```

JackTriton commented:
> Gotcha. Should that be "plain" texts?

Oh, sorry yes `PLAIN_TEXTS`

cjohnson57 reviewed (approved):
Thanks so much for all your changes! It's good now as far as I can tell. 

Before merging, we should also get approval from Fenhl, and probably more testing of full JP seeds.

JackTriton commented:
Add UnitTest that checks language file / properties

Below this is template for the errors

If the language file misses `property.json`
`{lang_name}: property.json is not included`

If some properties are missing / wrong in `property.json`
`{lang_name}: some properties are wrong in property.json`

If some bin files are not included in the `bin_patch.json`
`{lang_name}: some non-wanted bin files are included`

JackTriton commented:
Unless the conflicts revolve with in-game messages or custom texts, all conflicts will be resolved after merging

flagrama commented:
Conflicts _have_ to be fixed _before_ merging, that's just how git (and other version control software) works. Do you mean "after it is reviewed"?

fenhl commented:
Conflicts in generated files are fine to leave unresolved, they will be fixed by the maintainer who merges the PR.

flagrama commented:
> Conflicts in generated files are fine to leave unresolved, they will be fixed by the maintainer who merges the PR.

Oh, yeah, I'm on the app so don't know how to look at the actual conflicts. If that's all it is, ignore me.

fenhl commented on Dungeon.py:
```diff
@@ -3,17 +3,16 @@
 
 class Dungeon:
-    def __init__(self, world: World, name: str, hint: HintArea, regions: Optional[list[Region]] = None) -> None:
+    def __init__(self, world: World, name: str, hint: str, regions: Optional[list[Region]] = None) -> None:
```
I still think we shouldn't introduce language-specific code to these core business logic types. This means reverting the separation between `HintAreaDefault` and language-specific hint area types, and using the `HintArea` enum instead of a textual representation here.

JackTriton replied:
Instead, doing this wouldn't be confusing?:
Get language specific hint area, merge it with default area specification and use it as `HintArea`
That way separation won't be needed

fenhl replied:
`HintArea` is a core business logic type and shouldn't know about language specifics at all except in methods that generate text. I can send a PR if that helps?

JackTriton replied:
> I can send a PR if that helps?
That'll be great

JackTriton replied:
There's problem with `Region.py`
Current thing has something like below:
```
@property
    def hint(self) -> Optional[HintArea]:
        if self.hint_name is not None:
            return self.world.HintAreaLang[self.hint_name]
        if self.dungeon:
            return self.dungeon.hint

    @property
    def alt_hint(self) -> Optional[HintArea]:
        if self.alt_hint_name is not None:
            return self.world.HintAreaLang[self.alt_hint_name]
```
Because most of case the thing you want by using `at` is the ones for `HintAreaLang`, I still think this should use `HintAreaLang` instead of `HintAreaDefault`
Also, because `HintAreaDefault` is defined in `Hints.py`, it cannot import `HintAreaDefault`: it always gives uncertain error
What do you think?

JackTriton replied:
I'll just say resolved at this point with this update

fenhl commented on Hints.py:
```diff
@@ -350,25 +344,38 @@ def get_simple_hint_no_prefix(item: Item) -> Hint:
     text = gossip_text.text
     colors = list(gossip_text.colors) if gossip_text.colors is not None else []
     color = 'White'
+    if lang.base == "jp":
+        index = 0
+        while "#" in text[index:]:
+            pre_text = text[:index]
+            split_text = text[index:].split("#", 2)
+            split_text[0] = pre_text + split_text[0]
+            if not bool(re.match(r'^[0-9A-Fa-f]', split_text[1])):
+                if len(colors) > 0:
+                    color = colors.pop(0)
+                split_text[1] = COLOR_MAP[color][1] + split_text[1]
+                split_text[2] = "00" + split_text[2]
+            split_text[1] = "#" + split_text[1] + "#"
+            index += len(split_text[0]) + len(split_text[1])
+            text = "".join(split_text)
```
Can you add a comment explaining why this needs a separate code path for the JP text system? I know the escape codes are different but I don't see these being used here.

JackTriton replied:
In JP code, `#` is also used as signature for coloring by using `#XX`
This gets confused because hint also use `#` as universal coloring code
I'll add comment about why on the next commit

fenhl commented on Hints.py:
```diff
@@ -1687,236 +2199,324 @@ def build_altar_hints(world: World, messages: list[Message], include_rewards: bo
-        else:
-            text += char
-    return text
+    if lang != "jp":
```
Why are we skipping this for JP-base? Could at least use a comment pointing to where this is handled in that case, or maybe moving that code here.

JackTriton replied:
JP-base use those signatures on `Messages.py` as well
I'll add the comments regarding those

fenhl commented on Language.py:
```diff
@@ -0,0 +1,92 @@
+
+from Utils import lang_path
+
+def htz(s: str) -> str:
```
This needs a more descriptive name or a comment explaining what it does.

JackTriton replied:
Ok

fenhl commented on Patches.py:
```diff
@@ -1784,48 +1786,81 @@ def update_scrub_text(message: bytearray, text_replacement: list[str], default_p
+                    )
                 else:
-                    boss_location = next(filter(lambda loc: loc.type == 'Boss', world.get_entrance(f'{dungeon} Before Boss -> {dungeon.vanilla_boss_name} Boss Room').connected_region.locations))
+                    if world.settings.logic_rules == 'glitched':
```
This needs to be updated, glitched logic no longer needs special handling.

fenhl commented on SettingsList.py:
```diff
@@ -664,6 +673,15 @@ class SettingInfos:
         shared         = True,
     )
 
+    language_selection = Combobox(
```
This setting should be at the very top of the ROM Options tab, like on [my fork](https://ootrandomizer.com/generatorDev?version=devFenhl_). This applies to the setting definition here as well as the GUI definition in settings_mapping.

fenhl commented on World.py:
```diff
@@ -63,6 +64,19 @@ def __init__(self, world_id: int, settings: Settings, resolve_randomized_setting
+                if key in B:
+                    merged.update(B[key])
+                C[key] = merged
+            return C
```
The variable names should be in lowercase.
```suggestion
        def update_dict(a: dict, b: dict) -> dict:
            c = {}
            for key, subdict in a.items():
                merged = subdict.copy()
                if key in b:
                    merged.update(b[key])
                c[key] = merged
            return c
```

fenhl commented on language/.gitignore:
This file should be removed.

fenhl commented on language/English/property.json:
```diff
@@ -0,0 +1,9775 @@
+        ]
+    ],
+    "MISC_MESSAGES": [
+        [
+            50,
+            [
+                "\b\u0013\u0002You got \u0005ABombs\u0005@!\u0001If you see something\u0001suspicious, bomb it!",
+                35
+            ]
+        ],
```
These arrays should be objects instead, with keys that make it more obvious what the numbers mean. Same for many of the other fixed-length arrays in this file.

JackTriton replied:
Should I also use those signatures I implemented on JP-base to EN-base as well instead?
That way the code won't be visually confusing and know what's what

fenhl replied:
I'm sorry, I don't understand what you're referring to. Could you give an example?

JackTriton replied:
For example `\u0013\u0002` means `\x13\x02` for byte array notation
Because JSON cannot use byte array, this was auto-translated

On the other hand, JP-base uses specific codes instead
For `\x13\x02`, this can be translated as `★02`
That way language creator can avoid using array codes and wouldn't be confused by getting several arrays

Watching something like `\x13\x02\x05\x41` would be very confusing for us
Instead, `★02#01` would be better / universal I guess

fenhl replied:
I am not talking about control codes or string escapes. I'm referring to JSON arrays (`[]`) and objects (`{}`).

JackTriton replied:
Oh, that'll be much easier I guess

JackTriton replied:
Should I implement the same thing on the tables without numbers as well like `hintTable`?

fenhl replied:
Yes please.

JackTriton commented:
This update is to confirm what's good and not on the code

cjohnson57 commented:
Also, apologies, I just realized that we never added the "waiting for release" tag which is where this PR was for the last couple months.

JackTriton commented:
@cjohnson57 @fenhl In order to resolve all the conflicts, I have to make this branch up to date

Currently working on the text part and I think it'll be renewed in next 24 hours

JackTriton committed:
Part1

JackTriton committed:
Part2

JackTriton committed:
Part3 - Final

JackTriton committed:
Delete language/.DS_Store

JackTriton committed:
Delete language/__pycache__ directory

JackTriton committed:
Trailing for Python 3.9

JackTriton committed:
Typo fixes on Messages

JackTriton commented:
Currently, the branch conflicts revolving around ASM and generated data
The ASM part doesn't conflict with this branch so I think Pull request + rebuild would resolve the issue

cjohnson57 commented:
FYI, fenhl is preparing a PR to your branch for some hint region cleanup. In the meantime I will try to do a test seed with JP full randotext next week to see if anything breaks.

cjohnson57 commented:
FYI, tried to generate a seed on this branch using Japanese and got this error. The error did not occur when selecting English.

```
list index out of range
Traceback (most recent call last):
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\OoTRandomizer.py", line 57, in start
    main(settings)
    ~~~~^^^^^^^^^^
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 57, in main
    patch_and_output(settings, spoiler, rom)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 372, in patch_and_output
    patch_cosmetics_log = prepare_rom(spoiler, world, rom, settings, rng_state, restore_rom)
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Main.py", line 214, in prepare_rom
    patch_rom(spoiler, world, rom)
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Patches.py", line 1237, in patch_rom
    write_gossip_stone_hints(spoiler, world, messages)
    ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Hints.py", line 327, in write_gossip_stone_hints
    update_message_by_id(messages, id, str(gossip_text), lang, 0x23)
                                       ~~~^^^^^^^^^^^^^
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Hints.py", line 106, in __str__
    return get_raw_text(color_text(self, self.lang), self.lang.base)
                        ~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "C:\Users\Caleb\Repos\OoT-Randomizer-Tests\Hints.py", line 364, in color_text
    split_text[2] = "00" + split_text[2]
                           ~~~~~~~~~~^^^
IndexError: list index out of range
```

Using this settings string: `BSAWDNCAX2TBJYYGGAFSL62ANAWGEAAAAACAASKEASHAANLZAAADAKAEAEBABBXCAAABNAGJD8ASEJQ4VNFAGSCWEV43CAAAAAAAG2HPJNB2DBSSBUDWEARAAADACEJGUBC`

Changing the randotext setting did not help, but setting gossip stones to "no hints" did. Clearer hints or not did not affect it. Most hint distros did not help, however, setting it to "Useless" did fix the issue, which makes sense. Strangely, the error also did not occur with hints set to "Very Strong" or "Very Strong with Magic", but it did when set to "Strong", which would probably make sense if I had more intimate knowledge of the different hint distros.

Hopefully, the PR fenhl is preparing to address hints will resolve this issue.

cjohnson57 commented:
[OoT_A0FAE_GJAJIMJB4W_Spoiler.json](https://github.com/user-attachments/files/24632741/OoT_A0FAE_GJAJIMJB4W_Spoiler.json)

I started this seed, playing on PJ64, and got an immediate freeze when talking to the nearest NPC (Ingo)
<img width="1154" height="935" alt="image" src="https://github.com/user-attachments/assets/56f14fdd-38d5-4b8e-9707-ede9d304679d" />

Talking to other NPCs and reading signs isn't crashing, so I guess it's something specific to Ingo's textbox. Unfortunately I'm not sure how to check what randotext was assigned to him.

JackTriton commented:
> I messed around on 5 different seeds and here are my findings.
> 
> 
> 

Ok, this looks like the ones got shuffled didn't get the control code that was supposed to be attached (the "00" + glitch also comes from the same instance)

I'll use direct text output to see what caused those

JackTriton committed:
Goal place textboxes for World.py's Goal texts, Adding uninstant code before Goto ending to prevent freezes, replace '##' to '#' in hint texts

JackTriton commented:
Sorry there was another glitch so I have to fix it right now

JackTriton committed:
Color text fallout safes for JP texts

cjohnson57 commented:
Thanks, it seems to be working much better now. The hint patching is working and I didn't get any freezes immediately when talking to someone.

There's still the case where it doesn't let me advance the textbox, I'd guess this one's a more complicated fix...
<img width="1154" height="935" alt="image" src="https://github.com/user-attachments/assets/149fc7f5-2873-4aca-b998-dc9f4260a0c0" />

And then I found this case where it lets me click through the first 3 textboxes, but then the 4th one freezes for some reason:
<img width="1154" height="935" alt="image" src="https://github.com/user-attachments/assets/1fa1a159-a930-4625-83a9-385f1441066b" />
<img width="1154" height="935" alt="image" src="https://github.com/user-attachments/assets/1aea9224-90cf-46b7-bf24-eea70abe29e7" />

Also a couple textboxes that are weird but didn't cause any technical issues. I actually saw the 060B one twice, on two different seeds.
<img width="1154" height="935" alt="image" src="https://github.com/user-attachments/assets/94b78235-57b8-4951-bbd3-5f26b36582c4" />
<img width="1154" height="935" alt="image" src="https://github.com/user-attachments/assets/c90c2f6a-d7fc-486b-b4e2-e7ac7ce75c2f" />

cjohnson57 commented:
Btw, there's some code from @GSKirox that should be helpful for testing this PR, that adds a display on the screen for the current textbox ID. He's going to include it in a later PR that expands debug mode, but for now we can add it, then maybe just undo that commit before merging...

If you add the asm hook and C function for `message_id_check`, as well as the change in `ootSymbols.ld`, then it should work:
https://github.com/OoTRandomizer/OoT-Randomizer/pull/2493/files#diff-ec33c7eb10e1082a2d581ea448af77b53ce9d429f7aba1b5ae2ec98af4b430e3

JackTriton commented:
@cjohnson57 


> And then I found this case where it lets me click through the first 3 textboxes, but then the 4th one freezes for some reason
> 
Could you give me the setting seed / generation seed?
I'll look it up

> Also a couple textboxes that are weird but didn't cause any technical issues. I actually saw the 060B one twice, on two different seeds.
>
I searched through the file and found out that in JP rom, 060B and 016F has control codes as well
I'll edit the `is_id_message` to ignore the control codes and see what'll happen

JackTriton committed:
Various text fixes including preventing softlocks, pads on Compass / Map texts, unique id texts and reorganize Message.transform

JackTriton commented:
@cjohnson57 All of the presented issues are fixed as far as I know

fenhl reviewed (changes requested):
I posted a review in the form of PR https://github.com/JackTriton/OoT-Randomizer/pull/3.

fenhl committed:
Language review (#3)

* Revert addition of language-specific hint area types

* Add missing type annotations

* Reorder imports alphabetically

* Move generated file out of repo root and gitignore it

* Ensure files are read as UTF-8 on Windows

* Rename language_selection setting to language

* Move language dir into data dir

* Don't force-capitalize language display name

* Remove unused update_dict function

JackTriton committed:
Message: import HintArea, Patches: configure_dungeon_info fix, property.json: add shorter_name

JackTriton commented:
@fenhl Merged and edited some files in order to work it properly

1. Message
Added `import HintArea` for some of them in order to use it

2. Patches
Currently the system cannot use full-width characters so if the language cannot be decoded with ascii, it uses English's property instead for now

3. japanese/property.json
Added `shorter_name` property for future language updates

JackTriton committed:
Fixed some uncolored properties and line_wrap issue

fenhl committed:
Make language a shared setting (#5)

fenhl committed:
Unit tests for patching with Japanese language (#4)

fenhl commented:
Alright, @dotzo has volunteered to play a full seed with the language set to Japanese, and I'm sure @cjohnson57 will want to resume his text shuffle testing when he's back. I'm also curious how much in-game testing you've done on your own, that would help us identify what still needs to be tested.

JackTriton committed:
Some HintArea __str__ fix

JackTriton commented:
@fenhl Fixed

fenhl commented:
That's an interesting use case I hadn't thought of, but we can't simply revert it to non-shared because we want to show it in the spoiler log. There's also concerns regarding racing if you can reroll the seed with arbitrary language changes, I'll bring it up for discussion internally.

fenhl commented:
I don't think grouping language with cosmetics is feasible since it affects so much at patch time and we need to keep backward and forward compatibility for everything for cosmetics.

cjohnson57 commented:
Hello, I'm back from vacation and testing some more.

There's definitely been massive improvement! I tested easily 150+ textboxes across 4 different seeds including NPCs, gossip stones, items, and shops, and only found 2 problem textboxes. Neither were hard crashes, just softlocks where I was unable to advance the text.

<img width="1154" height="935" alt="image" src="https://github.com/user-attachments/assets/2c248236-942a-4ce9-af98-47c00b13df04" />
<img width="1154" height="935" alt="image" src="https://github.com/user-attachments/assets/8c1fa30c-6bd3-4672-9922-268898e81707" />

JackTriton committed:
Message fix for some ceratin dialogues

JackTriton commented:
@cjohnson57 For much convenient solution, I think implementing debugging system that dumps texts during patching would be great
Using Zelda's Letter tool on Cloudmodding would also be great solution for checking texts

JackTriton committed:
Message debug, Reparse the align, Reconfigure some grammatical errors

JackTriton commented:
@cjohnson57 @flagrama  
All of the presented issues are now fixed

JackTriton committed:
Fix / Enhance Ocarina note issue

JackTriton commented:
@cjohnson57 Fixed + implemented new system  
Texts like `[A]` and `[C Up]` will automatically translated into codes like `0x839F` and `0x83A5` in Japanese  
This doesn't affect on the `display` and `get_string` for keeping smoothness of the debugging

JackTriton committed:
Fix for Plain text, minor get_string edit for better control with control codes

JackTriton committed:
Added useful functions to the Language, change gender in dungeon text into dungeon_gender, added notes for both formatting and textures

JackTriton committed:
Add language specific replace table to exsisting language properties

JackTriton committed:
Fix area text issue

fenhl commented:
Sorry, between multiple maintainers being on vacation and a critical issue coming up that needed to be handled before the upcoming 9.1 release, there hasn't been much time spent on this. As far as testing is concerned, this PR is ready to go, and the racing impact has been deemed acceptable, but we haven't reached a consensus on how it should interact with the settings string or how it should be presented in the GUI.

JackTriton commented:
> > Weird that there is no reference to this in decomp from what I can see. However it does seem like rando has it correct for the way it was implemented in game to me too.
> 
> I don't think the actual accented characters would be listed anywhere in the codebase, just the character IDs corresponding to textures, which wouldn't be present in decomp

@cjohnson57 
I think I know what's happening
I searched through the decompiled file for both NTSC and GC EU MQ Debug and found out that for `gMsgChar81LatinCapitalLetterIWithCircumflexTex.i4`, `Á` is listed as its texture on NTSC rom but for GC EU MQ Debug rom however, the listed character was `Î` instead so I think they already patched the problem for NTSC roms

JackTriton committed:
Make the language as cosmetic, add entry 0x9097
